File: apps/backend/app/models/execution.py
from sqlmodel import SQLModel, Field, Relationship, Column, JSON
from typing import Optional, List, Dict, Any
from datetime import datetime

class WorkflowExecution(SQLModel, table=True):
    """Tracks each workflow execution instance"""
    __tablename__ = "workflow_executions"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    flow_version_id: int = Field(foreign_key="flow_versions.id")
    conversation_id: Optional[int] = Field(default=None, foreign_key="conversations.id")
    
    # Execution status
    status: str  # 'running', 'completed', 'failed', 'cancelled'
    started_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: Optional[datetime] = None
    
    # Data
    input_data: Dict[str, Any] = Field(default_factory=dict, sa_column=Column(JSON))
    output_data: Optional[Dict[str, Any]] = Field(default=None, sa_column=Column(JSON))
    error_message: Optional[str] = None
    
    # Progress tracking
    total_nodes: int = 0
    completed_nodes: int = 0
    
    # Relationships
    flow_version: "FlowVersion" = Relationship(back_populates="executions")
    node_executions: List["NodeExecution"] = Relationship(back_populates="workflow_execution")


class NodeExecution(SQLModel, table=True):
    """Tracks individual node execution within a workflow"""
    __tablename__ = "node_executions"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    workflow_execution_id: int = Field(foreign_key="workflow_executions.id")
    
    # Node info
    node_id: str  # ID from the flow JSON
    node_type: str  # 'start', 'message', 'ai-reply', etc.
    node_label: str
    
    # Execution status
    status: str  # 'pending', 'running', 'completed', 'failed', 'skipped'
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    execution_time_ms: Optional[int] = None
    
    # Data
    input_data: Dict[str, Any] = Field(default_factory=dict, sa_column=Column(JSON))
    output_data: Optional[Dict[str, Any]] = Field(default=None, sa_column=Column(JSON))
    error_message: Optional[str] = None
    
    # Relationships
    workflow_execution: WorkflowExecution = Relationship(back_populates="node_executions")

File: apps/web/lib/types/execution.ts

export interface Execution {
  id: string | number
  flowId?: number
  flow_version_id?: number
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  startedAt?: string
  started_at?: string
  completedAt?: string
  completed_at?: string
  duration?: number
  duration_ms?: number
  error?: string
  error_message?: string
  input?: any
  output?: any
  nodeExecutions?: NodeExecution[]
  total_nodes?: number
  completed_nodes?: number
  success_rate?: number
}

export interface NodeExecution {
  id: string
  nodeId: string
  nodeName: string
  nodeType: string
  status: 'pending' | 'running' | 'success' | 'failed' | 'skipped'
  startedAt?: string
  completedAt?: string
  duration?: number
  input?: any
  output?: any
  error?: string
}

export interface ExecutionUpdate {
  executionId: string
  flowId: string
  status: 'running' | 'completed' | 'failed'
  nodeId?: string
  nodeStatus?: 'running' | 'success' | 'failed'
  output?: any
  error?: string
}

export interface TimelineEvent {
  id: string
  status: 'completed' | 'failed' | 'running' | 'pending'
  timestamp: string
  nodeName: string
  duration?: number
}

export interface ExecutionStatusBadgeProps {
  status: 'running' | 'completed' | 'failed' | 'cancelled'
}

export interface ExecutionTimelineProps {
  events: TimelineEvent[]
}

export interface NodeExecutionCardProps {
  execution: {
    id: string
    nodeName: string
    status: 'success' | 'failed' | 'running'
    duration: number
    input?: any
    output?: any
    error?: string
  }
}

